### ISSUE TITLE
Question on custom models

### ISSUE THREAD

**Issue by vince62s** [BlackSamorez/tensor_parallel#88]
Hi,
without using transformers / accelerate blablabla, what are the constraints on the model to be tensor paralelizable ?

does it need to be a nn.Sequential ? does input dimensions need to be always in the same order ?

I am trying to load a model on two gpus but only the first is being allocated. (both are visible)

**Comment by BlackSamorez** [comment]
Hi @vince62s !
For a model to be tensor-paralleliazable automatically it should consist of basic _Linear/Convolution/Embedding PyTorch_ layers. Other basic layers will still work but they will be replicated on each _GPU_. The problem with autoconfig may arise in runtime if those basic layers' weights are accessed directly (e.g. with linear.weight) since those accesses are not taken into account when planning device communications.
Otherwise, everything should generally work, especially when deploying the model, since it should deal with any _PyTorch_ parameters and buffers no problem. If it's not the case for you, feel free to submit a more detailed report on your problem and I'll take a look.

**Comment by vince62s** [comment]
ok, let me explain a bit.
When you use bitsandbytes to quantize (8it or 4bit) you need to load data in a layer and then call cuda() which triggers the quantization. Also you cannot call cuda() twice otherwise it screws up everything.
To overcome cpu ram usage when loading a big model here is what I do.
I build an empty model in cpu ram, then module by module I load data and move it to gpu module by module.
here is where it happens:
https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/models/model.py#L71-L92
now, I tried to use tensor_parallel at each module level (by modifying this code) and it only uses one gpu.

**Comment by BlackSamorez** [comment]
I'll have a more detailed look later, but right now you can try and take a look at how quantized deployment is done in [tensor_parallel int4 demo on kaggle](https://www.kaggle.com/code/blacksamorez/tensor-parallel-int4-llm/).

**Comment by BlackSamorez** [comment]
Are you quantizing the model before making it `tensor_parallel`? Because that wouldn't work. What you need to do is to create a model on `CPU`/`meta`, then make it tensor_parallel and only then dispatch it on GPUs (quantization happens during dispatch).
Here's a simplified example of how I did it for the [Kaggle int4 demo](https://www.kaggle.com/code/blacksamorez/tensor-parallel-int4-llm):
```python
import torch
import transformers
import accelerate
from transformers.utils.bitsandbytes import replace_with_bnb_linear

import tensor_parallel as tp

# Initialize a model on meta device (zero allocations)
with accelerate.init_empty_weights():
    model = transformers.AutoModelForCausalLM.from_config(...)

# Make it tensor_parallel
model = tp.TensorParallelPreTrainedModel(model)

# Mark linear layers to be quantized
model = replace_with_bnb_linear(
    model,
    quantization_config=transformers.utils.quantization_config.BitsAndBytesConfig(...),
)

# Load, dispatch and quantize the weights
device_map = tp.infer_sharded_device_map(model) # infer where to put each weight
for <EACH CHECKPOINT>: 
    converted_state_dict = tp.convert_state_dict( # <- tensor_parallel helper function. 
        torch.load(<CHECKPOINT PATH>),            #    Creates a tensor_parallel checkpoint form a normal one
        ...
    )
        
    # Dispatch the checkpoint
    for param_name, param in converted_state_dict.items():
        module_name = param_name
        while len(module_name) > 0 and module_name not in device_map:
            module_name = ".".join(module_name.split(".")[:-1])
        accelerate.utils.set_module_tensor_to_device(model, param_name, device_map[module_name], value=param)
```

From what I see, you're trying to use an already written dispatch code (from within `load_state_dict`). It probably will not work with `tensor_parallel` since it doesn't convert the `state_dict` and (obviously) doesn't check for internal `tensor_parallel` wrapper prefixes.
Please try and use `accelerate` to dispatch the model as shown in the _int4 demo_. If you have any questions I'll be happy to help!

**Comment by vince62s** [comment]
I'll try but then you confirm that this line `model = tp.TensorParallelPreTrainedModel(model)` does not move anything to gpu, right ?

**Comment by BlackSamorez** [comment]
It does, unless the model is on `meta` device. Then it remains on `meta` device.

**Comment by vince62s** [comment]
so it HAS to be on meta otherwise quantization won't work in the snippet above, am I correct ?

**Comment by vince62s** [comment]
I am not seeing such print out. I tried various things as you mentioned but no luck.
can't figure out how to make this work.

**Comment by BlackSamorez** [comment]
Firstly, about `tp.infer_sharded_device_map`: it has nothing to do with `tp.Sharded`. I should probably rename it. What it does is that it simply creates a mapping of `parameter name->parameter device`. I created this function to simplify [deployment with accelerate](https://github.com/huggingface/accelerate/blob/535b52cef22c81a5cd01f971d0a573dd1c73762d/src/accelerate/utils/modeling.py#L994).

Secondly, about `tp.Sharded`: when trainable parameters are replicated on multiple GPUs with tensor parallelism (e.g. LayernNorm weights can't be split) we average them once in a while for them not to diverge. To do this, `tp.TensorParallel` can be wrapped with `tp.Sharded` with `model = tp.Sharded(model)`. This needs to be done after model weights are loaded in and after they are properly frozen/unfrozen (to decide which weights need averaging). But to wrap tp.TensorParallelPreTrainedModel you need to do `model.wrapped_model = tp.Sharded(model.wrapped_model)` cause reasons. It's a hot mess and I was hoping to refactor it soon.

**tldr**: call `tp.tensor_parallel(..., sharded=False)` between 1) and 2). In 3) convert state dicts with `tp.convert_state_dict` and use `tp.infer_sharded_device_map` to determine on which GPU to put resulting weights. After 4) use `tp.Sharded` as described above.

**Comment by vince62s** [comment]
but the issue is `tp.infer_sharded_device_map` works when being called before the bnb replacement (getting the mappign fine), but after I am getting the same error as above:
raise KeyError(
KeyError: "Can't decide where to put {name} in a sharded model state dict. Are you sure it's a sharded dict?"

**Comment by BlackSamorez** [comment]
I'll merge a PR fixing this message (and possibly a problem behind it) today. Then I'll be able to tell what goes wrong there.

**Comment by vince62s** [comment]
well to be sure I just tested the kaggle notebook and to me it is still very unclear what is going on:
I am loading as a test the decapoda/llama-7b-hf
when printing the device_map, it returns all params duplicated on each device, for instance:
'wrapped_model.module_shards.0.model.layers.0.self_attn.tp_wrapped_module.q_proj.weight': device(type='cuda', index=0)
'wrapped_model.module_shards.1.model.layers.0.self_attn.tp_wrapped_module.q_proj.weight': device(type='cuda', index=1)

after the state_dict load both GPU are loaded with 13-14GB which means each of them carries the full model.

even more trouble some, in 4-bit I would expect the model footprint to be 4GB, split over two GPUs, hence 2GB each

**Comment by BlackSamorez** [comment]
>'wrapped_model.module_shards.0.model.layers.0.self_attn.tp_wrapped_module.q_proj.weight': device(type='cuda', index=0)
'wrapped_model.module_shards.1.model.layers.0.self_attn.tp_wrapped_module.q_proj.weight': device(type='cuda', index=1)

Those are `module_shards.0` and `module_shards.1`. They are supposed to have identical structure since those are two parts of the same model put on two devices.
As for 13-14GB for `decapoda/llama-7b-hf`: this should not be the case. [The original demo](https://www.kaggle.com/code/blacksamorez/tensor-parallel-int4-llm) showcases `decapoda-research/llama-30b-hf` which wouldn't physically fit if it wasn't properly quantized and split between two GPUs.

**Comment by BlackSamorez** [comment]
By the way [v1.2.6](https://github.com/BlackSamorez/tensor_parallel/releases/tag/v1.2.6) has just been released which fixes the error message discussed above and, hopefully, some other aspects of dispatch.

**Comment by vince62s** [comment]
> Those are module_shards.0 and module_shards.1. They are supposed to have identical structure since those are two parts of the same model put on two devices.
Then I don't undertsand anything. How is the model sharded on the 2 devices? layer N going to device 0 and layer M going to device 1? or part of layer N going device 0 and otehr part of layer N going to device 1 ?
I don't undertsand based on this device_map how things are dispatched.

**Comment by vince62s** [comment]
> As for 13-14GB for decapoda/llama-7b-hf: this should not be the case. [The original demo](https://www.kaggle.com/code/blacksamorez/tensor-parallel-int4-llm) showcases decapoda-research/llama-30b-hf which wouldn't physically fit if it wasn't properly quantized and split between two GPUs.

I updated the `accelerate` library and the footprint is ok now. so I need to see how to replicate this with my code.

**Comment by BlackSamorez** [comment]
> > As for 13-14GB for decapoda/llama-7b-hf: this should not be the case. [The original demo](https://www.kaggle.com/code/blacksamorez/tensor-parallel-int4-llm) showcases decapoda-research/llama-30b-hf which wouldn't physically fit if it wasn't properly quantized and split between two GPUs.
> 
> I updated the `accelerate` library and the footprint is ok now. so I need to see how to replicate this with my code.

Yes, fp4 support for dispatch was [added](https://github.com/huggingface/accelerate/pull/1505) only like a few weeks ago.

**Comment by BlackSamorez** [comment]
> > Those are module_shards.0 and module_shards.1. They are supposed to have identical structure since those are two parts of the same model put on two devices.
> 
> Then I don't undertsand anything. How is the model sharded on the 2 devices? layer N going to device 0 and layer M going to device 1? or part of layer N going device 0 and otehr part of layer N going to device 1 ? I don't undertsand based on this device_map how things are dispatched.

The entire point of this library is that each layer is split individually and each GPU has only a portion of each layer. This way all of the *GPUs* can be utilized simultaneously when using the model.

**Comment by vince62s** [comment]
I think I'm not far but I have the feeling that the auto config does not work for  my model (remember it is not "PreTrained") so maybe I may have to make up a config.

**Comment by vince62s** [comment]
On another note, would that be possible to have the same behavior with a model on "cpu" as on "meta".
If we build an empty model on "cpu" (with the skipt_init function) then we could also call tp.tensor_parallel() without actually moving to cuda and dispatch manually after that. The difference would be to avoid using that hacky method set_module_tensor_to_device(), we could just use param.to(device) it works.

**Comment by BlackSamorez** [comment]
> On another note, would that be possible to have the same behavior with a model on "cpu" as on "meta".
> If we build an empty model on "cpu" (with the skipt_init function) then we could also call tp.tensor_parallel() without actually moving to cuda and dispatch manually after that. The difference would be to avoid using that hacky method set_module_tensor_to_device(), we could just use param.to(device) it works.

That could come in handy. I'll see what I can do.
